### PR TITLE
RPC: Refactor to share common code between examples

### PR DESCRIPTION
**Pull request by rgoliver**
#### Problem
As more examples get RPC support it is clear that a lot of code is unnecessarily duplicated. 

#### Change overview

- Move the common RPC services into a common location and have default virtual implementations which use the ember API.
-  Applications can customize the behavior when necessary by overriding the methods.
- Also cleanup the lighting proto to better match the spec.

#### Testing
- Tested the RPC version of the lighting app on EFR NRF and Linux and ensured Device, Lighting, and Button RPC's worked as expected.
- Tested the ESP RPC locking app compiled with RPCs (Don't have the correct board to test functionality)
- Tested the ESP RPC all-clusters app and verified Device, Lighting, and Button RPC's worked as expected.

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

File: examples/lighting-app/linux/BUILD.gn [hunk 1]
```diff
@@ -57,6 +57,11 @@ executable("chip-lighting-app") {
       "Rpc.cpp",
     ]
 
+    include_dirs += [
```
Comment (mspang) [inline examples/lighting-app/linux/BUILD.gn:60]: Please do not directly add include_dirs for your dependencies.

**Review by mspang — CHANGES_REQUESTED** [review]

File: examples/common/pigweed/rpc_services/Lighting.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            response.color.hue        = hue;
+            response.color.saturation = saturation;
+            response.has_color        = true;
+        }
+
+        return pw::OkStatus();
+    }
+
+protected:
+    bool _support_level;
```
[the remark this answers is not in the document]
Comment (rgoliver) [reply examples/common/pigweed/rpc_services/Lighting.h]: done

File: examples/common/pigweed/rpc_services/Locking.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+namespace chip {
+namespace rpc {
+
+class Locking final : public generated::Locking<Locking>
+{
+
+public:
+    virtual ~Locking() = default;
+
+    virtual pw::Status Set(ServerContext &, const chip_rpc_LockingState & request, pw_protobuf_Empty & response)
```
[the remark this answers is not in the document]
Comment (rgoliver) [reply examples/common/pigweed/rpc_services/Locking.h:36]: These are the top level common virtual implementations which can be optionally overriden by the applications if needed.

File: examples/lighting-app/linux/BUILD.gn (hunk 1 above)
Comment (rgoliver) [reply examples/lighting-app/linux/BUILD.gn]: Done

File: examples/common/pigweed/rpc_services/Lighting.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        uint8_t on = request.on;
+        RETURN_STATUS_IF_NOT_OK(
+            emberAfWriteServerAttribute(1, ZCL_LEVEL_CONTROL_CLUSTER_ID, ZCL_ON_OFF_ATTRIBUTE_ID, &on, ZCL_BOOLEAN_ATTRIBUTE_ID));
+
+        if (_support_level && request.has_level)
+        {
+            // Clip level to max
+            uint8_t level = std::min(request.level, static_cast<uint32_t>(std::numeric_limits<uint8_t>::max()));
+            RETURN_STATUS_IF_NOT_OK(emberAfWriteServerAttribute(1, ZCL_LEVEL_CONTROL_CLUSTER_ID, ZCL_CURRENT_LEVEL_ATTRIBUTE_ID,
```
[the remark this answers is not in the document]
Comment (rgoliver) [reply examples/common/pigweed/rpc_services/Lighting.h]: Done

**Review by woody-apple — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Pull request merged**